File: vscode-extension/src/rules/supabase/rlsPolicySuggestionRule.ts
import { Rule } from '../rule';
import { Finding, FindingCategory, FindingConfidence } from '../../models/finding';
import { ProjectContext, suggestedPolicyForTable } from '../../scanner/projectContext';

export class RlsPolicySuggestionRule implements Rule {
  readonly code = 'rls-policy-suggestion';

  evaluate(context: ProjectContext): Finding[] {
    const findings: Finding[] = [];
    const seenTables = new Set<string>();

    for (const access of context.tableAccesses) {
      const normalized = access.table.toLowerCase();
      if (!seenTables.has(normalized)) { seenTables.add(normalized); } else { continue; }

      const confidence = this._confidenceForTable(normalized);
      if (!confidence) { continue; }

      const policy = suggestedPolicyForTable(access.table);
      if (!policy) { continue; }

      findings.push(new Finding({
        category: FindingCategory.suggestion,
        confidence,
        code: this.code,
        message: `Heuristic RLS suggestion for table '${access.table}' (${confidence.toUpperCase()} confidence — verify against your schema)`,
        fix: `Consider: \`${policy}\`  — this is a heuristic; confirm column names match your actual schema before applying.`,
        filePath: access.file.relativePath,
        line: access.line,
      }));
    }
    return findings;
  }

  private _confidenceForTable(normalized: string): FindingConfidence | null {
    switch (normalized) {
      case 'profiles':
      case 'users':
        return FindingConfidence.high;
      case 'posts':
      case 'todos':
      case 'notes':
      case 'orders':
      case 'comments':
      case 'messages':
        return FindingConfidence.medium;
      default:
        return null;
    }
  }
}

File: lib/src/rules/supabase/rls_policy_suggestion_rule.dart
import '../../models/finding.dart';
import '../../models/project_context.dart';
import '../../rule.dart';

/// Emits heuristic RLS policy suggestions for tables the app accesses.
///
/// Suggestions are only shown for tables with HIGH or MEDIUM confidence
/// mappings (i.e. well-known tables in [suggestedPolicyForTable]).
/// Unknown tables get no suggestion — a wrong suggestion is worse than silence.
class RlsPolicySuggestionRule extends Rule {
  const RlsPolicySuggestionRule();

  @override
  String get code => 'rls-policy-suggestion';

  @override
  List<Finding> evaluate(ProjectContext context) {
    final findings = <Finding>[];
    final seenTables = <String>{};

    for (final access in context.tableAccesses) {
      final normalized = access.table.toLowerCase();
      if (!seenTables.add(normalized)) {
        continue;
      }

      final confidence = _confidenceForTable(context, normalized);

      // Only emit for HIGH or MEDIUM confidence tables — skip unknowns.
      if (confidence == null) {
        continue;
      }

      final policy = suggestedPolicyForTable(
        access.table,
        ddl: context.ddlMetadata,
      );
      if (policy == null) {
        continue;
      }

      findings.add(
        Finding(
          category: FindingCategory.suggestion,
          confidence: confidence,
          code: code,
          message:
              "Heuristic RLS suggestion for table '${access.table}' "
              '(${confidence.label} confidence — verify against your schema)',
          fix:
              'Consider: `$policy`  — this is a heuristic; confirm column names match your actual schema before applying.',
          filePath: access.file.relativePath,
          line: access.line,
        ),
      );
    }

    return findings;
  }

  /// Returns confidence level for a table we know about, or null for unknowns.
  ///
  /// Tier 1 — DDL-resolved (FK to `auth.users`): MEDIUM. Concrete schema
  /// evidence ties the table to user-owned data, but we still don't know
  /// the policy semantics match without runtime data.
  ///
  /// Tier 2 — hardcoded HIGH/MEDIUM ladder for the canonical Supabase tables.
  ///
  /// Returns null for tables that are neither DDL-declared nor in the
  /// hardcoded list — no suggestion is emitted (a wrong suggestion is worse
  /// than silence).
  FindingConfidence? _confidenceForTable(
    ProjectContext context,
    String normalized,
  ) {
    if (context.ddlMetadata.forTable(normalized) != null) {
      return FindingConfidence.medium;
    }
    switch (normalized) {
      // HIGH: canonical Supabase auth tables — column names are prescribed.
      case 'profiles':
      case 'users':
        return FindingConfidence.high;
      // MEDIUM: common user-owned content tables with conventional columns.
      case 'posts':
      case 'todos':
      case 'notes':
      case 'orders':
      case 'comments':
      case 'messages':
        return FindingConfidence.medium;
      default:
        return null; // unknown table → no suggestion
    }
  }
}
